### DISCUSSION TITLE
Call only specified functions

### DISCUSSION THREAD

**Discussion by simetelum** [KeyserDSoze/Rystem.OpenAi#35] [Ideas]
I couldn't find that feature but it would be great if we were able to specify which functions can be called automatically.
So far, if we call `request.ExecuteAsync(true)` - it would call any registered function that met some conditions. 
I don't know if my explanation is good, but I'm looking for something like this:  
`request.WithFunction(ExampleFunction.NameLabel, true)`
where boolean value will determine if function will be called automatically or not.

**Comment by KeyserDSoze** [comment]
If I understood well, you may try with an extension method to add the behavior to ChatRequestBuilder.

```
public static class ChatRequestBuilderExtensions
{
    public static ChatRequestBuilder IfTrueWithFunction(this ChatRequestBuilder builder, string name, bool condition)
    {
        if (condition)
            return builder.WithFunction(name);
        return builder;
    }
}
```

So you can choose which one you want to use in this way
```
var response = await openAiApi.Chat
                .RequestWithUserMessage("What is the weather like in Boston?")
                .WithModel(ChatModelType.Gpt35Turbo_Snapshot)
                .IfTrueWithFunction(WeatherFunction.NameLabel, true)
                .ExecuteAndCalculateCostAsync(true);
```

**Reply by simetelum** [reply]
Thank you for your reply, but that is not what i need. I am registering my functions with DI and I want all my functions to be included within request. The thing is, if I execute request as `request.ExecuteAsync(true)` , code execution continues in `WrapAsync(string message)` method of defined function, and it doesn't matter if I return `null` or new `FunctionResponseModel` object from `WrapAsync(string message)` method, there is always going to be an extra request to OpenAI API - this is main issue for me. Also, code that called `request.ExecuteAsync(true)` will not recieve response where `FinishReason` is `function` but `stop`. 

I do not know if I explained it well, or maybe it is not event doable right now, but thanks anyway.

**Comment by KeyserDSoze** [comment]
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/102e0938-2754-4583-a99d-72f39f923135)

![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/9c22e89f-3205-479f-af1a-3102ca637aae)

![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/7b99841d-bc54-4877-b8cf-3ef19dc19057)

![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/bb6dda5b-58d2-457c-9b35-9ae372b4cfcd)

![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/23bbea36-fa4e-47bd-85a0-269fcad26f5a)

Now I'm checking the null reference returned from WrapAsync to avoid the further request. Furthermore, now you have as finish reason "null" to understand that is a finish reason for null in function and in addition I left the argument and name of the function.

**Reply by simetelum** [reply]
Thank you very much, that is exactly what I needed.
